Bugfix gdb.debug: exe parameter now respected

goreil opened the pull request:
Reopening #2227, since the Coverage tests didn't seem to handle changing the target branch correctly.

This commit now properly supports the exe parameter in `pwnlib/gdb.py:debug()`, allowing a different argv[0] than the executable. (See Issue #1273 )

It achieves this by leveraging the gdbsever`--wrapper` argument with a python script that calls execve with the specified args.
For remote connections `ssh.process(run=False)` already creates a python script that we can use.


## Details
1. We create a python script, that calls execve with our arguments
`/tmp/pwn-XXXXXX.py`
```python
#!<local python version>
import ctypes
...
ctypes.CDLL(None).execve(exe, argv, env)
```

2. `gdbserver` has no native support for `argv[0]`

Therefore we leverage the `--wrapper` flag, which allows us to execute an arbitrary command. `gdbserver` will attach to the first unforked `execve` system call.

The command will then be
`gdbserver --wrapper /tmp/pwn-XXXXXX.py -- <other gdbserver args>`

goreil committed:
Bugfix gdb.debug: exe parameter now respected

This commit now properly supports the exe parameter in
`pwnlib/gdb.py:debug()`, allowing a different argv[0] than the
executable.

It achieves this by leveraging the gdbsever`--wrapper` argument
with a python script that calls execve with the specified args.

goreil committed:
Merge branch 'dev' into stable-gdb-exe

goreil committed:
Backintegration for python2

goreil committed:
Update Changelog.md

goreil committed:
pwnlib gdb.py, try to pass these tests...

goreil committed:
Encode script in ssh.process(run=False) for tests

goreil committed:
Re-trigger Pull request tests

goreil committed:
Merge branch 'Gallopsled:dev' into stable-gdb-exe

goreil commented:
Hi, just wanted to ask if there are any changes you require me to make to the pull request or is it ready to merge?

peace-maker commented:
I didn't have time to test this yet :/

peace-maker commented on pwnlib/gdb.py:
That is already handled by `misc.normalize_argv_env` before calling this helper.

goreil replied:
Thank you for pointing it out!
`misc.normalize_argv_env` does not encode bytearrays, i.e 
```py
misc.normalize_argv_env([bytearray(b'b')], {}, log)
Out[2]: ([bytearray(b'b')], {})
```
This is an issue, since `ctypes.c_char_p` only works with `bytes` and not `bytearrays`.
```py
In [14]: ctypes.c_char_p(bytearray(b'a'))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[14], line 1
----> 1 ctypes.c_char_p(bytearray(b'a'))

TypeError: bytes or integer address expected instead of bytearray instance
In [15]: ctypes.c_char_p(b'a')
Out[15]: c_char_p(140315614962176)
```

Therefore this line is needed for the code to work. I tested it already by deleting the line and then the code breaks.

peace-maker replied:
If it's just the `bytearray`, you can drop the `packing._encode` call.

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -296,7 +357,10 @@ def _gdbserver_args(pid=None, path=None, args=None, which=None, env=None):
+    elif env is not None:
         env_args = []
         for key in tuple(env):
             if key.startswith(b'LD_'): # LD_PRELOAD / LD_LIBRARY_PATH etc.
```
This appears to strip any `LD_*` environment variables from the environment in which `gdbserver` is executed and only passes them to the target program. We should mimic this in the execve wrapper.

goreil replied:
I wrote a new testcase in ` gdb.py: Add test case for LD_Preload` showin that LD_Preload is still parsed correctly.

peace-maker replied:
Are you sure gdbserver isn't run with the copied libc as well there? The test doesn't tell. We only want the target binary to run with the modified libraries.

goreil replied:
Alright, it's in the new commit now! Thanks for pointing it out!

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -529,12 +614,21 @@ def debug(args, gdbscript=None, exe=None, ssh=None, env=None, sysroot=None, api=
     if env:
         env = {bytes(k): bytes(v) for k, v in env}
 
+    exe = which(packing._decode(exe or args[0]))
```
There's a similar check below around line 655, so this might be a duplicate/interfering? The other code doesn't check for existence on ssh backends, can you dig into why and possibly mirror this here?

goreil replied:
Good catch, thanks!

So the check in line 655 calls `os.path.exists` which doesn't make sense in the SSH-Environment. But I'm unsure why `os.path.exists` is called in the first place, since thats already the job of the `which` call.

I checked the [commit](https://github.com/Gallopsled/pwntools/commit/f727b66475dc8eb779e2d0af56250892f7642fa9) + the related Pull Request:  #1323 but didn't find any explanation for the code.

Therefore I'm just going to delete the duplicate check.

peace-maker replied:
`os.path.exists` is there to check if the path passed in the `exe` parameter exists. Yes it's redundant when `exe` is None and it's grabbing `args[0]` using which. Now you're calling `which` on the `exe` parameter too.

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -529,12 +614,21 @@ def debug(args, gdbscript=None, exe=None, ssh=None, env=None, sysroot=None, api=
+        # but can use the ``--wrapper`` option to execute commands and catches
+        # ``execve`` calls.
+        # Therefore, we use a wrapper script to execute the target binary
+        script = _execve_script(args, executable=exe, env=env, ssh=ssh)
```
Can we only use this wrapper workaround if `exe != args[0]` to save some complexity for the usual case?

goreil replied:
Added in new commit `gdb.py: easier script if argv[0] == exe`

peace-maker commented on pwnlib/gdb.py:
Maybe add a check that `key` does not contain a `=`. That is usually done by python's os.execve.

goreil replied:
Added in new commit `Add check that "=" not in misc.normalize_argv_env`

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -458,6 +522,13 @@ def debug(args, gdbscript=None, exe=None, ssh=None, env=None, sysroot=None, api=
         >>> io.close()
+        
+        Start a new process with modified argv[0]
+        >>> io = gdb.debug("/bin/sh", 'continue', argv=[b'\xde\xad\xbe\xef'])
```
This usage is weird, since `gdb.debug` doesn't have an `argv` parameter but only an `args` one and this appears to rely on `process` having an `argv` parameter? I'd expect this to look like `gdb.debug([b'\xde\xad\xbe\xef'], 'continue', exe="/bin/sh")`. Same for the other added examples.

I'm confused on how this works with `gdbserver = runner(args, env=env, aslr=1, **kwargs)` when `runner` is `tubes.process.process`. The first argument is argv and the `kwargs` contain the explicit `argv` argument again, no? Which one is used?

goreil replied:
Fixed in commit `gdb.py: Add test case for LD_Preload`

peace-maker replied:
No, there are still calls of `gdb.debug` with an `argv` argument. There is no `argv`, only `args` in `gdb.debug`.

peace-maker reviewed (changes requested):
Sorry for the huge delay!

goreil committed:
gdb.py: easier script if argv[0] == exe

goreil committed:
Add check that "=" not in misc.normalize_argv_env

This check checks prevents the use of "=" in the
key of an environment variable, which is generally
impossible.

goreil committed:
gdb.py: Add test case for LD_Preload

goreil committed:
gdb.py Correct handling of LD_ env-variables

peace-maker committed:
Merge branch 'dev' into stable-gdb-exe

peace-maker commented on pwnlib/gdb.py:
Can you add the same things as the other ssh.process PR? Use explicit `ctypes.CDLL('libc.so.6')` and handle execve failing.

goreil replied:
Sure! I implemented the changes in latest commit.

peace-maker commented on pwnlib/gdb.py:
Let's be consistent and use `prefix='pwnlib-execve-'` like the ssh module does.
Also can you use a `with tempfile.NamedTemporaryFile(..) as tmp:` context here please?

goreil replied:
Sure! I implemented the changes in latest commit.

peace-maker replied:
You missed this instance of the same code 🚓

goreil replied:
Oh, I changed the wrong one. Nice catch 🦹‍♂️
Fixed now

peace-maker commented on pwnlib/gdb.py:
```diff
@@ -296,13 +357,19 @@ def _gdbserver_args(pid=None, path=None, args=None, which=None, env=None):
+    
+    if python_wrapper_script is not None:
+        gdbserver_args += ['--wrapper', python_wrapper_script, '--']
+    else:
```
```suggestion
    elif env is not None:
```

goreil committed:
Merge branch 'Gallopsled:dev' into stable-gdb-exe

goreil committed:
Update pwnlib/gdb.py

Co-authored-by: peace-maker <peacemakerctf@gmail.com>

goreil committed:
gdb.py address comments

1. explicit ctypes.CDLL('libc.so.6'), handle execve failing
2. consistent namedTempFile
3. drop packing._encode() since it's done earlier
4. testcases solve argv-args confusion

peace-maker commented:
Please have a look at the test failures here: https://github.com/Gallopsled/pwntools/actions/runs/6259779607/job/16996463774?pr=2233#step:16:303

goreil committed:
gdb.py: Fix Namedtempfile-prefix + Bugfix

goreil committed:
Restore prefix for gdbscript

goreil commented:
Ok, that passes the tests now.

goreil committed:
Merge branch 'dev' into stable-gdb-exe

Arusekk commented:
Actually the wrapper script could (and maybe should?) be reused with spawning carefully tailored remote processes over ssh, we already have a big Python function there. I think that function is not the best piece of code currently, but having two such functions is too many probably.

goreil committed:
Unify execve wrapper script under one function

goreil committed:
Merge branch 'stable-gdb-exe' of github.com:Youheng-Lue/pwntools into stable-gdb-exe

goreil committed:
Merge branch 'dev' into stable-gdb-exe

goreil committed:
gdb.py, Remove leftover script

goreil commented:
I addressed the wrapper script now by creating a new function in 
`misc.create_execve_script` 

Anything else that needs to be resolved?

peace-maker committed:
Fix logging scope and ignore_environ argument

peace-maker committed:
Merge branch 'dev' into stable-gdb-exe

peace-maker reviewed (approved):
Very cool, thank you for your persistence working on this!

I've changed the few things I noticed myself to speed this up.

Ori-Pe commented:
@goreil @peace-maker

Tried to set arg0 to another value and debugging does not work, is it supposed to work today?
Tested in latest pwntools version (4.15.0, https://github.com/Gallopsled/pwntools/tree/4.15.0).
Tested also in the latest dev version (dev5.0.0, https://github.com/Gallopsled/pwntools/tree/dev).

Snippet:
`io = gdb.debug(exe='./mychallenge', args=['fake_arg0'], api=True)`

<img width="1218" height="227" alt="image" src="https://github.com/user-attachments/assets/c2a175e1-b8ce-4af8-90e2-845aa6a5d0d3" />
(No process is running under gdb, and the args are not set if I try to run again).

(Tried to set the arch manually and no success, I do not think it's really related).

I can debug with edited args when I debug the "python3 exp.py" (Which internally is os.execve(challenge, modified_args, env) but it is not easy to work with.

My gdb.debug function and gdb overall are working well, just this situation.
Running on VMware Ubuntu24.04


EDIT: https://github.com/Gallopsled/pwntools/pull/2691
